### PR TITLE
ENH: HDF5-based FX Reader/Writer.

### PR DESCRIPTION
**Pull request by ssanderson**
HDF5-based implementation of `FXRateReader`.

- Adds new classes, `HDF5FXRateReader` and `HDF5FXRateWriter` for reading/writing fx rates stored in hdf5.
- Moves existing content of `zipline/data/fx.py` into a directory. We now have `zipline.data.fx` as a module, with child modules implementing readers, plus `zipline.data.fx.base` defining the interface they all implement.
- Adds a basic test suite that should be suitable for testing any implementation of `FXRateReader`. 
  - Adds subclasses for testing `InMemoryFXRateReader` (previously only indirectly tested via the pipeline API tests) and the new `HDF5FXRateReader`.
  - Updates testing infrastructure to be able to share code between in memory and h5-based tests.

### CODE REVIEW

File: zipline/data/fx/in_memory.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        self._data = data
+        self._default_rate = default_rate
+
+    def get_rates(self, rate, quote, bases, dts):
+        """Get rates to convert ``bases`` into ``quote``.
+        """
+        if rate == 'default':
+            rate = self._default_rate
+
+        df = self._data[rate][quote]
```
Comment (quantophred) [inline zipline/data/fx/in_memory.py]: Looks like calling `.get_rates(bogus_rate, valid_quote ...)` returns the exact same KeyError from the exact same line as `.get_rates(valid_rate, bogus_quote, ...)`. Is it worth splitting this line into two lines, to help distinguish them? (Maybe even using different KeyError subclasses for the two? Seems like that would encourage weird control flow in calling code, though.)

File: zipline/data/fx/in_memory.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        #
+        # Logically, the operation here is:
+        #
+        # (df
+        #  .reindex(dts, side='right')
+        #  .reindex_axis(cols, axis='columns')
+        #  .values)
+        #
+        # But pandas' performance on the above is not great, and we call this
+        # method a lot, so we implement our own indexing logic.
```
Comment (quantophred) [inline zipline/data/fx/in_memory.py]: Is it worth adding tests to make sure the two operations are identical?

File: zipline/data/fx/hdf5.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+        # There aren't many columns in the output array, so it's easier to pull
+        # all columns and reindex in memory. For rows, however, a quick and
+        # easy optimization is to pull just the slice from min(row_ixs) to
+        # max(row_ixs).
+        min_row = row_ixs[0]
+        max_row = row_ixs[-1]
+        rows = dataset[min_row:max_row + 1]  # +1 to be inclusive of end
+
+        return rows[row_ixs - min_row][:, col_ixs]
```
Comment (quantophred) [inline zipline/data/fx/hdf5.py]: I'm having trouble keeping track of what the rows and cols are. Worth renaming `row` to `date`, and `col` to `currency`? (Are those interpretations actually correct?)

File: zipline/data/fx/exploding.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class ExplodingFXRateReader(implements(FXRateReader)):
+    """An FXRateReader that raises an error when used.
+
+    This is useful for testing contexts where FX rates aren't actually needed.
+    """
+
+    def get_rates(self, rate, quote, bases, dts):
+        raise AssertionError("FX rates requested unexpectedly!")
```
[the remark this answers is not in the document]
Comment (ssanderson) [reply zipline/data/fx/exploding.py]: Talked about this in person, but making this a default would prevent `interface` from being able to catch typos in the method name in an implementation.

File: zipline/data/fx/hdf5.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+    def get_rates(self, rate, quote, bases, dts):
+        """Get rates to convert ``bases`` into ``quote``.
+        """
+        if rate == 'default':
+            rate = self._default_rate
+
+        self._check_dts(self.dts, dts)
+
+        row_ixs = self.dts.searchsorted(dts, side='right') - 1
```
[the remark this answers is not in the document]
Comment (ssanderson) [reply zipline/data/fx/hdf5.py]: Clarified this here: https://github.com/quantopian/zipline/pull/2597/files#diff-018677a08bc4e4c73e81c2d2e2bf4402R27-R28 .

**Pull request merged**